In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
#print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.

In [3]:
# Read data file
data = pd.read_csv('../Assignment/Reviews.csv')

In [4]:
# Have a look at the data
data.head(3)

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   

                 Summary                                               Text  
0  Good Quality Dog Food  I have bought several of the Vitality canned d...  
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...  
2  "Delight" says it all  This is a confection that has been around a fe...

In [5]:
# We will only consider the 'Score' and the 'Text' column only for this classification problem
df = data[['Score', 'Text']]

In [6]:
# Taking a look at the modified dataframe
df.head(20)

Score                                               Text
0       5  I have bought several of the Vitality canned d...
1       1  Product arrived labeled as Jumbo Salted Peanut...
2       4  This is a confection that has been around a fe...
3       2  If you are looking for the secret ingredient i...
4       5  Great taffy at a great price.  There was a wid...
5       4  I got a wild hair for taffy and ordered this f...
6       5  This saltwater taffy had great flavors and was...
7       5  This taffy is so good.  It is very soft and ch...
8       5  Right now I'm mostly just sprouting this so my...
9       5  This is a very healthy dog food. Good for thei...
10      5  I don't know if it's the cactus or the tequila...
11      5  One of my boys needed to lose some weight and ...
12      1  My cats have been happily eating Felidae Plati...
13      4  good flavor! these came securely packed... the...
14      5  The Strawberry Twizzlers are my guilty pleasur...
15      5  My daughter loves twizzlers and this shipment ...
16      2  I love eating them and they are good for watch...
17      5  I am very satisfied with my Twizzler purchase....
18      5  Twizzlers, Strawberry my childhood favorite ca...
19      5  Candy was delivered very fast and was purchase...

In [7]:
df.shape

(568454, 2)

In [8]:
# Looking at how many reviews have '3' rating
df.loc[df["Score"]==3].count()

Score    42640
Text     42640
dtype: int64

In [9]:
# Excluding all those reviews having '3' rating
df1 = df.loc[df['Score']!= 3]

In [10]:
df1.shape

(525814, 2)

In [11]:
# Adding a new column called 'Classified_score' where 1 and 2 rating is mapped as 0 (negative) and 4, 5 rating is mapped as 1 (positive)
df1['Classified_score'] = [0 if x < 3 else 1 for x in df1['Score']]

F:\Anaconda\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [12]:
# Having a look at the modified data
df1.head(20)

Score                                               Text  Classified_score
0       5  I have bought several of the Vitality canned d...                 1
1       1  Product arrived labeled as Jumbo Salted Peanut...                 0
2       4  This is a confection that has been around a fe...                 1
3       2  If you are looking for the secret ingredient i...                 0
4       5  Great taffy at a great price.  There was a wid...                 1
5       4  I got a wild hair for taffy and ordered this f...                 1
6       5  This saltwater taffy had great flavors and was...                 1
7       5  This taffy is so good.  It is very soft and ch...                 1
8       5  Right now I'm mostly just sprouting this so my...                 1
9       5  This is a very healthy dog food. Good for thei...                 1
10      5  I don't know if it's the cactus or the tequila...                 1
11      5  One of my boys needed to lose some weight and ...                 1
12      1  My cats have been happily eating Felidae Plati...                 0
13      4  good flavor! these came securely packed... the...                 1
14      5  The Strawberry Twizzlers are my guilty pleasur...                 1
15      5  My daughter loves twizzlers and this shipment ...                 1
16      2  I love eating them and they are good for watch...                 0
17      5  I am very satisfied with my Twizzler purchase....                 1
18      5  Twizzlers, Strawberry my childhood favorite ca...                 1
19      5  Candy was delivered very fast and was purchase...                 1

In [13]:
# We will split our data into training and test dataset
X = df1['Text'] # feature
y = df1['Classified_score'] # Target

In [14]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y)

In [15]:
# Now, first we need to process the text and then choose between classifiers for best performance
# We will use Tf-Idf vectorizer to transform the text into vectorised form to apply ML algorithms

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.naive_bayes import MultinomialNB

classification = Pipeline([('tfidf', TfidfVectorizer()), ('clf', MultinomialNB())])

classification = classification.fit(X_train, y_train)

In [16]:
# we need to check, how well this classifier worked

y_predicted = classification.predict(X_test)
np.mean(y_predicted == y_test)

0.8636329058073546

# Our naive bayes classifier has a accuracy of 86.36 %, which is quite good. We would still look at some other classifiers



In [17]:
from sklearn.svm import LinearSVC

clasf_svm = Pipeline([('tfidf', TfidfVectorizer()),('svm', LinearSVC())])

clasf_svm = clasf_svm.fit(X_train, y_train)

In [18]:
y_predicted_svm = clasf_svm.predict(X_test)
np.mean(y_predicted_svm == y_test)

0.9458137447319975

# Linear SVC gives us an accuracy of 94.58% 

In [19]:
from sklearn.linear_model import SGDClassifier
clasf_sgd = Pipeline([('tfidf', TfidfVectorizer()),('sgd', SGDClassifier())])

clasf_sgd = clasf_sgd.fit(X_train, y_train)

F:\Anaconda\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDClassifier'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


In [20]:
y_predicted_sgd = clasf_sgd.predict(X_test)
np.mean(y_predicted_sgd == y_test)

0.911406271395317

# SGD classifier gives us an accuracy of 91.14% 

# We can further tune the hyperparameters by applying Grid-search to increase the accuracy.



# From our observation, we conclude, Linear Support Vector Classifier would be the best model to predict the rating (positive or negative)